### PR TITLE
Adding changes to add the Pegasus Onnx Config.

### LINKED ISSUE #16308: ONNXConfig: Add a configuration for all available models
This issue is about the working group specially created for this task. If you are interested in helping out, take a look at this [organization](https://huggingface.co/OWG), or add me on Discord: `ChainYo#3610`

We want to contribute to HuggingFace's ONNX implementation for all available models on HF's hub. There are already a lot of architectures implemented for converting PyTorch models to ONNX, but we need more! We need them all!

Feel free to join us in this adventure! Join the org by clicking [here](https://huggingface.co/organizations/OWG/share/TskjfGaGjGnMXXssbPPXrQWEIbosGqZshZ)

Here is a non-exhaustive list of models that all models available:
- [x] Albert
- [x] BART
- [x] BeiT
- [x] BERT
- [x] BigBird
- [x] BigBirdPegasus
- [x] Blenderbot
- [x] BlenderbotSmall
- [x] BLOOM
- [x] CamemBERT
- [ ] CANINE
- [x] CLIP
- [x] CodeGen
- [x] ConvNext
- [x] ConvBert
- [ ] CTRL
- [ ] CvT
- [x] Data2VecText
- [x] Data2VecVision
- [x] Deberta
- [x] DebertaV2
- [x] DeiT
- [ ] DecisionTransformer
- [x] DETR
- [x] Distilbert
- [ ] DPR
- [ ] DPT
- [x] ELECTRA
- [ ] FNet
- [ ] FSMT
- [x] Flaubert
- [ ] FLAVA
- [ ] Funnel Transformer
- [ ] GLPN
- [x] GPT2
- [x] GPTJ
- [x] GPT-Neo
- [ ] GPT-NeoX
- [ ] Hubert
- [x] I-Bert
- [ ] ImageGPT
- [ ] LED
- [x] LayoutLM
- [ ] 🛠️ LayoutLMv2
- [x] LayoutLMv3
- [ ] LayoutXLM
- [ ] LED
- [x] LeViT
- [x] Longformer
- [x] LongT5
- [ ] 🛠️ Luke
- [ ] Lxmert
- [x] M2M100
- [ ] MaskFormer
- [x] mBart
- [ ] MCTCT
- [ ] MPNet
- [x] MT5
- [x] MarianMT
- [ ] MegatronBert
- [x] MobileBert
- [x] MobileViT
- [ ] Nyströmformer
- [x] OpenAIGPT-2
- [ ] 🛠️ OPT
- [x] OWLViT
- [x] PLBart
- [ ] Pegasus
- [x] Perceiver
- [ ] PoolFormer
- [ ] ProphetNet
- [ ] QDQBERT
- [ ] RAG
- [ ] REALM
- [ ] 🛠️ Reformer
- [ ] RemBert
- [x] ResNet
- [ ] RegNet
- [ ] RetriBert
- [x] RoFormer
- [x] RoBERTa
- [ ] SEW
- [ ] SEW-D
- [ ] SegFormer
- [ ] Speech2Text
- [ ] Speech2Text2
- [ ] Splinter
- [x] SqueezeBERT
- [ ] Swin Transformer
- [x] T5
- [ ] TAPAS
- [ ] TAPEX
- [ ] Transformer XL
- [x] TrOCR
- [ ] UniSpeech
- [ ] UniSpeech-SAT
- [ ] VAN
- [x] ViT
- [ ] Vilt
- [ ] VisualBERT
- [ ] Wav2Vec2
- [ ] WavLM
- [ ] XGLM
- [x] XLM
- [ ] XLMProphetNet
- [x] XLM-RoBERTa
- [x] XLM-RoBERTa-XL
- [ ] 🛠️ XLNet
- [x] YOLOS
- [ ] Yoso

🛠️ next to a model suggests that the PR is in progress. If there is nothing next to a model, it means that ONNX does not yet support the model, and thus we need to add support for it.

If you need help implementing an unsupported model, here is a [guide](https://huggingface.co/docs/transformers/serialization#exporting-a-model-for-an-unsupported-architecture) from HuggingFace's documentation.

If you want an example of implementation, I did one for [CamemBERT ](https://github.com/huggingface/transformers/pull/14059) months ago.

### PR DESCRIPTION
**Pull request by pramodith**
# What does this PR do?
This pull request makes the required changes to support running the Pegasus model in Onnx runtime, Linked to 
A related PR #18305 was closed because of some git issues I was running into and replaced for this one.

Fixes  https://github.com/huggingface/transformers/issues/16308

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [X] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. Linked to https://github.com/huggingface/transformers/issues/16308
- [X] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [X] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@lewtun 
@ChainYo

### CODE REVIEW

**Comment by pramodith** [comment]
@lewtun I tested the export and validated the onnx model using the following code.

```
from transformers.models.pegasus.configuration_pegasus import PegasusOnnxConfig, PegasusConfig
from transformers.models.pegasus import PegasusModel, PegasusTokenizer, PegasusForConditionalGeneration
from transformers.onnx import export, validate_model_outputs
from pathlib import Path

def check_onnx_model(task):
    if task == "default":
        config = PegasusConfig.from_pretrained("google/pegasus-x-base")
        model = PegasusModel.from_pretrained("google/pegasus-x-base")
        tokenizer = PegasusTokenizer.from_pretrained("google/pegasus-x-base")
    elif task == "seq2seq-lm":
        config = PegasusConfig.from_pretrained("google/pegasus-xsum")
        model = PegasusForConditionalGeneration.from_pretrained("google/pegasus-xsum", add_cross_attention=True)
        tokenizer = PegasusTokenizer.from_pretrained("google/pegasus-xsum")
    else:
        config = PegasusConfig.from_pretrained("google/pegasus-xsum")
        model = PegasusForConditionalGeneration.from_pretrained("google/pegasus-xsum")
        tokenizer = PegasusTokenizer.from_pretrained("google/pegasus-xsum")

    onnx_config = PegasusOnnxConfig(config, task=task, use_past=True)
    onnx_path = Path("model.onnx")
    onnx_inputs, onnx_outputs = export(tokenizer, model, onnx_config, onnx_config.default_onnx_opset, onnx_path)
    print(onnx_inputs)
    print(onnx_outputs)
    print(validate_model_outputs(onnx_config,tokenizer,model,onnx_path,onnx_outputs,onnx_config.atol_for_validation))

check_onnx_model("seq2seq-lm")
```

For the case of "seq2seq-lm" I got the following error when I set `use_past = True`. `ValueError: Outputs values doesn't match between reference model and ONNX exported model: Got max absolute difference of: 0.00010585784912109375
`
Is the difference in magnitude of order 1e-4 acceptable?

**Comment by ChainYo** [comment]
> For the case of "seq2seq-lm" I got the following error when I set `use_past = True`. `ValueError: Outputs values don't match between the reference model and ONNX exported model: Got max absolute difference of: 0.00010585784912109375 ` Is the difference in magnitude of order 1e-4 acceptable?

Depending on models, 1e-3 is acceptable. I wouldn't go further.

File: src/transformers/models/pegasus/configuration_pegasus.py [hunk 1]
```diff
@@ -167,3 +172,227 @@ def num_attention_heads(self) -> int:
     @property
     def hidden_size(self) -> int:
         return self.d_model
+
+
+# Copied from transformers.models.bart.configuration_bart.BartOnnxConfig with Bart->Pegasus
+class PegasusOnnxConfig(OnnxSeq2SeqConfigWithPast):
```
Comment (lewtun) [inline src/transformers/models/pegasus/configuration_pegasus.py:178]: You can deal with the tolerance issue by adding the following property to this class (you could try with 1e-4 first and lower to 1e-3 if needed):

```
    @property
    def atol_for_validation(self) -> float:
        return 1e-4
```

**Review by lewtun — APPROVED** [review]
Thanks for adding ONNX support of this architecture @pramodith 🔥 !

The PR is very clean and I've left a small suggestion to tweak the tolerance level. Once you've included this, could you confirm the slow tests pass with:

```
RUN_SLOW=1 pytest tests/onnx/test_onnx_v2.py -k "pegasus"
```

**Comment by lewtun** [comment]
> Thanks for adding ONNX support of this architecture @pramodith 🔥 !
> 
> The PR is very clean and I've left a small suggestion to tweak the tolerance level. Once you've included this, could you confirm the slow tests pass with:
> 
> ```
> RUN_SLOW=1 pytest tests/onnx/test_onnx_v2.py -k "pegasus"
> ```

Hey @pramodith just checking if you were able to run the slow tests successfully?

File: src/transformers/models/pegasus/configuration_pegasus.py (hunk 1 above)
Comment (pramodith) [reply src/transformers/models/pegasus/configuration_pegasus.py:178]: @lewtun sorry, just got to checking on this. I'm seeing the below when I run the tests. Does the **F** mean failed? I tried with both 1e-3 and 1e-4 and I see that if I don't use the RUN_SLOW flag the tests seem to succeed.
image

**Pull request closed without merging**